### PR TITLE
Fix Cannot fetch certificate: no endpoints available for service "http:sealed-secrets-controller:"

### PR DESCRIPTION
**Pull request by sathieu**
See https://istio.io/latest/docs/ops/configuration/traffic-management/protocol-selection/#explicit-protocol-selection

- fixes https://github.com/bitnami-labs/sealed-secrets/issues/502
- fixes https://github.com/bitnami-labs/sealed-secrets/issues/694

### CODE REVIEW

**Comment by sathieu** [comment]
@juan131 Please review :pray:.

**Comment by juan131** [comment]
hanks so much for your contribution @sathieu!!

We're performing a major refactorization of the chart at https://github.com/bitnami-labs/sealed-secrets/pull/687
Let's wait until this is done before continue with your changes

**Comment by juan131** [comment]
@sathieu the new major version was just released! 
```
$ helm repo update sealed-secrets
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "sealed-secrets" chart repository
Update Complete. ⎈Happy Helming!⎈
$  helm search repo sealed-secrets/sealed-secrets
NAME                         	CHART VERSION	APP VERSION	DESCRIPTION
sealed-secrets/sealed-secrets	2.0.0        	v0.17.1    	Helm chart for the sealed-secrets controller.
```
Please feel to rebase this branch in your fork with the latest changes in the main branch, and update this PR.

File: cmd/kubeseal/main.go [hunk 1]
```diff
@@ -51,24 +51,25 @@ const (
...
-	sealingScope   ssv1alpha1.SealingScope
-	reEncrypt      bool // re-encrypt command
-	unseal         = flag.Bool("recovery-unseal", false, "Decrypt a sealed secrets file obtained from stdin, using the private key passed with --recovery-private-key. Intended to be used in disaster recovery mode.")
-	privKeys       = flag.StringSlice("recovery-private-key", nil, "Private key filename used by the --recovery-unseal command. Multiple files accepted either via comma separated list or by repetition of the flag. Either PEM encoded private keys or a backup of a json/yaml encoded k8s sealed-secret controller secret (and v1.List) are accepted. ")
+	certURL            = flag.String("cert", "", "Certificate / public key file/URL to use for encryption. Overrides --controller-*")
+	controllerNs       = flag.String("controller-namespace", metav1.NamespaceSystem, "Namespace of sealed-secrets controller.")
+	controllerName     = flag.String("controller-name", "sealed-secrets-controller", "Name of sealed-secrets controller.")
+	controllerPortName = flag.String("controller-port-name", "", "Port name of sealed-secrets controller.")
```
Comment (juan131) [inline cmd/kubeseal/main.go:57]: It seems this parameter isn't used in the `main()` function when invoking the `run` function... Also.. It seems we're already assuming that the SealedSecret service has a port with name "http" in functions such as `validateSealedSecret` or `reEncryptSealedSecret`, see:

- https://github.com/bitnami-labs/sealed-secrets/blob/main/cmd/kubeseal/main.go#L316
- https://github.com/bitnami-labs/sealed-secrets/blob/main/cmd/kubeseal/main.go#L350

Couldn't do a similar assumption in `openCertCluster`?

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks so much for the PR @sathieu !! Please check my comments

File: helm/sealed-secrets/templates/deployment.yaml [hunk 2]
```diff
@@ -45,6 +45,7 @@ spec:
           args:
             - --key-prefix
             - {{ .Values.secretName | quote }}
+            - --controller-port-name=http
```
Comment (juan131) [inline helm/sealed-secrets/templates/deployment.yaml:48]: Please bump the patch version in the [Chart.yaml](https://github.com/bitnami-labs/sealed-secrets/blob/main/helm/sealed-secrets/Chart.yaml)

**Review by juan131 — CHANGES_REQUESTED** [review]

File: helm/sealed-secrets/templates/deployment.yaml (hunk 2 above)
Comment (juan131) [reply helm/sealed-secrets/templates/deployment.yaml:48]: Follow up question: you're implementing this flag for `kubeseal` but the `controller` ignores it. Am I right?

**Comment by sathieu** [comment]
@juan131 Comments addressed. Please review again.

**Comment by andoshin11** [comment]
This is indeed critical one. Appreciate the fix.

**Comment by sathieu** [comment]
@juan131 Anything missing in this PR?

**Review by juan131 — APPROVED** [review]
LGTM! I can confirm this fix works

**Comment by sathieu** [comment]
@juan131 Maybe this fix deserve a major version bump?

Also the CI is failing, is this related to this change?

**Review by juan131 — CHANGES_REQUESTED** [review]
@sathieu I just realized that this fixes the chart.. but it breaks the compatibility with the `controller.yaml` manifests that we publish at:

- https://github.com/bitnami-labs/sealed-secrets/releases/tag/v0.17.1

These manifests don't have a "http" name in the service port.

> maybe this fix deserve a major version bump?

Why do you think so? I don't see that they break backwards compatibility.

**Comment by juan131** [comment]
@sathieu I updated your branch with a suggestion that dynamically obtains the svc port name. Please let me know what you think

**Review by alvneiayu — APPROVED** [review]
Awesome work, LGTM

**Comment by sathieu** [comment]
@juan131 Thanks.  This is great! lgtm!

**Comment by linkvt** [comment]
Hello @juan131 , I'm addressing this here because I think your change broke the sealed-secrets-controller for k8s users with no permissions in kube-system - no blame intended.

As you can see in this line the proxier role only grants permissions to get `services/proxy` and not `services` https://github.com/bitnami-labs/sealed-secrets/blob/1490884d3b0d64c27900e63abfc0b268d49eaab2/helm/sealed-secrets/templates/role.yaml#L43-L45

I guess this is because `Services(namespace).ProxyGet` was replaced with `Services(namespace).Get`. Was there a reason for this? Otherwise I could create a PR to add `- services` to the Roles resources field to make it work again.

Thanks for checking.

**Comment by juan131** [comment]
Hi @linkvt 

You're totally right!! That should be pretty straight forward to fix since it just requires to update the RBAC permissions. I'll send a PR to fix it.

> Was there a reason for this?

We still use a proxy to access the service (I didn't change that), but previously we access the services to obtain the service port name information.

**Comment by chrisob** [comment]
@juan131 Coming off of @linkvt 's comment, is there no alternative that would support using the original `Services(namespace).ProxyGet`? Reason being, just simply allowing `get services` in RBAC as is the case in #715 punches a bigger hole in RBAC which may not be desired in multitenant environments (especially considering *any* authenticated user can get the entire service's manifest!).

Maybe I'm a little over cautious, but I feel that only allowing `services/proxy` RBAC was a nice solution which prevented random users from discovering more information about sealed-secrets-controller than they need, a-la the principle of least privilege.

**Comment by chrisob** [comment]
Just brainstorming here, but one approach could be to control this "discover service endpoint" functionality via feature flag, but by default have kubeseal assume the default `http` port & not attempt any service lookup.

**Comment by juan131** [comment]
Hi @chrisob 

Thanks for sharing your concerns, we really appreciate your opinion.  IMHO, the security risk is not that high. 

It limits the scope to the namespace where the SealedSecrets controller is installed, offering the opportunity to use a dedicated namespace for the controller so authenticated users can exclusively obtain the controller service metadata. Furthermore, this service is by default a "ClusterIP" service that exposes a single HTTP port, I don't think we add much value in terms of security by hiding this info from authenticated users.

**Comment by mkmik** [comment]
My plan for fixing this was to create a ConfigMap with the public key 
and avoid all this mess with actively trying to talking to the controller.

**Pull request merged**